packages/firestore/test/unit/specs/spec_test_runner.ts

```diff
@@ -1089,21 +1111,122 @@ class MockDocument {
   }
 }
 
+/**
+ * `WebStorage` mock that implements the WebStorage API used by Firestore.
+ */
+class MockStorage {
+  private static sharedData = new Map<string, string>();
+  private static activeInstances = new Map<EventListener, MockStorage>();
+
+  eventListener: EventListener | null = null;
+
+  getItem(key: string): string | null {
+    if (MockStorage.sharedData.has(key)) {
+      return MockStorage.sharedData.get(key);
+    }
+    return null;
+  }
+
+  removeItem(key: string): void {
+    const oldValue = MockStorage.sharedData.get(key);
+    MockStorage.sharedData.delete(key);
+    this.raiseStorageEvent(key, oldValue, null);
+  }
+
+  setItem(key: string, value: string): void {
+    const oldValue = MockStorage.sharedData.get(key);
+    MockStorage.sharedData.set(key, value);
+    this.raiseStorageEvent(key, oldValue, value);
+  }
+
+  raiseStorageEvent(
+    key: string,
+    oldValue: string | null,
+    newValue: string | null
+  ): void {
+    MockStorage.activeInstances.forEach((storage, listener) => {
+      // WebStorage doesn't raise events for writes from the originating client.
+      if (storage !== this) {
+        listener({
+          key: key,
+          oldValue: oldValue,
+          newValue: newValue,
+          storageArea: storage
+        } as any);
+      }
+    });
+  }
+
+  start(eventListener: EventListener): void {
+    MockStorage.activeInstances.set(eventListener, this);
+    this.eventListener = eventListener;
+  }
+
+  shutdown(): void {
+    MockStorage.activeInstances.delete(this.eventListener);
+    this.eventListener = null;
+  }
+}
+
+/**
+ * `Window` mock that implements the event and storage API that is used by
+ * Firestore.
+ */
+class MockWindow {
+  private webStorage = new MockStorage();
+
+  get localStorage(): Storage {
+    return this.webStorage as any;
```

Comment:
Doesn't tslint complain?  FWIW it's probably not hard to make MockStorage fully implement Storage but I don't care either way.